kdc/httpkadmind.c

```diff
@@ -0,0 +1,2025 @@
+    }
+    if (r && r->hcontext) {
+        heim_audit_addreason((heim_svc_req_desc)r, "%s", formatted);
+        audit_trail(r, code);
+    }
+    krb5_free_error_message(context, k5msg);
+
+    if (ret == -1 || msg == NULL) {
+        if (context)
+            krb5_log_msg(context, logfac, 1, NULL, "Out of memory");
+        return resp(r, MHD_HTTP_SERVICE_UNAVAILABLE,
+                    MHD_RESPMEM_PERSISTENT, NULL,
+                    "Out of memory", sizeof("Out of memory") - 1, NULL, NULL);
+    }
+
+    ret = resp(r, http_status_code, MHD_RESPMEM_MUST_COPY,
+               NULL, msg, strlen(msg), NULL, NULL);
+    free(formatted);
+    free(msg);
+    return ret == -1 ? -1 : code;
+}
+
+static krb5_error_code
+bad_req(kadmin_request_desc r,
+        krb5_error_code code,
+        int http_status_code,
+        const char *fmt,
+        ...)
+{
+    krb5_error_code ret;
+    va_list ap;
+
+    va_start(ap, fmt);
+    ret = bad_reqv(r, code, http_status_code, fmt, ap);
+    va_end(ap);
+    return ret;
+}
+
+static krb5_error_code
+bad_enomem(kadmin_request_desc r, krb5_error_code ret)
+{
+    return bad_req(r, ret, MHD_HTTP_SERVICE_UNAVAILABLE,
+                   "Out of memory");
+}
+
+static krb5_error_code
+bad_400(kadmin_request_desc r, int ret, const char *reason)
+{
+    return bad_req(r, ret, MHD_HTTP_BAD_REQUEST, "%s", reason);
+}
+
+static krb5_error_code
+bad_401(kadmin_request_desc r, const char *reason)
+{
+    return bad_req(r, EACCES, MHD_HTTP_UNAUTHORIZED, "%s", reason);
+}
+
+static krb5_error_code
+bad_403(kadmin_request_desc r, krb5_error_code ret, const char *reason)
+{
+    return bad_req(r, ret, MHD_HTTP_FORBIDDEN, "%s", reason);
+}
+
+static krb5_error_code
+bad_404(kadmin_request_desc r, const char *name)
+{
+    return bad_req(r, ENOENT, MHD_HTTP_NOT_FOUND,
+                   "Resource not found: %s", name);
+}
+
+static krb5_error_code
+bad_405(kadmin_request_desc r, const char *method)
+{
+    return bad_req(r, EPERM, MHD_HTTP_METHOD_NOT_ALLOWED,
+                   "Method not supported: %s", method);
+}
+
+static krb5_error_code
+bad_method_want_POST(kadmin_request_desc r)
+{
+    return bad_req(r, EPERM, MHD_HTTP_METHOD_NOT_ALLOWED,
+                   "Use POST for making changes to principals");
+}
+
+#if 0
+static krb5_error_code
+bad_500(kadmin_request_desc r,
+        krb5_error_code ret,
+        const char *reason)
+{
+    return bad_req(r, ret, MHD_HTTP_INTERNAL_SERVER_ERROR,
+                   "Internal error: %s", reason);
+}
+#endif
+
+static krb5_error_code
+bad_503(kadmin_request_desc r,
+        krb5_error_code ret,
+        const char *reason)
+{
+    return bad_req(r, ret, MHD_HTTP_SERVICE_UNAVAILABLE,
+                   "Service unavailable: %s", reason);
+}
+
+static krb5_error_code
+good_ext_keytab(kadmin_request_desc r)
+{
+    krb5_error_code ret;
+    size_t bodylen;
+    void *body;
+    char *p;
+
+    if (!r->keytab_name || !(p = strchr(r->keytab_name, ':')))
```

Comment:
Check that the file name is really in the temp dir that we're using so that we don't accidentally send any other file.